clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -103,6 +103,12 @@ public class HeartbeatRequestManager implements RequestManager {
      */
     private final BackgroundEventHandler backgroundEventHandler;
 
+    /**
+     * Timer for tracking the time since the last consumer poll.  If the timer expires, the consumer will stop
+     * sending heartbeat until the next poll.
+     */
```

Comment:
well, i'm not sure actually.  According to KIP-848: `Static membership, introduced in KIP-345, is still supported by this new rebalance protocol. When a member wants to leave temporary – e.g. while being bounced – it should send an heartbeat with a member epoch equals to -2. This signals to the group coordinator that the member left but will rejoin within the session timeout. When the member rejoins with the same instance ID, the group coordinator replaces the old member by the new member and gives back its current assignment.`

Reply:
I think you meant : only when `group.instance.id` is not set, i.e. dynamic member?

This is the current implementation
```
 // Starting from 2.3, only dynamic members will send LeaveGroupRequest to the broker,
        // consumer with valid group.instance.id is viewed as static member that never sends LeaveGroup,
        // and the membership expiration is only controlled by session timeout.
        if (isDynamicMember() && !coordinatorUnknown() &&
            state != MemberState.UNJOINED && generation.hasMemberId()) {
            // this is a minimal effort attempt to leave the group. we do not
            // attempt any resending if the request fails or times out.
            log.info("Member {} sending LeaveGroup request to coordinator {} due to {}",
                generation.memberId, coordinator, leaveReason);
            LeaveGroupRequest.Builder request = new LeaveGroupRequest.Builder(
                rebalanceConfig.groupId,
                Collections.singletonList(new MemberIdentity().setMemberId(generation.memberId).setReason(JoinGroupRequest.maybeTruncateReason(leaveReason)))
            );

            future = client.send(coordinator, request).compose(new LeaveGroupResponseHandler(generation));
            client.pollNoWakeup();
        }
```

Reply:
My take is that we should always send the leave group when the poll timer expires (group.instance.id null or not). Dynamic members should send -1 and static -2 as described in the KIP. The [documentation of the max.poll.interval](https://kafka.apache.org/documentation/#consumerconfigs_max.poll.interval.ms) is just describing what will happen on the broker side (if static member leaving, broker won't re-assign the partitions until the session timeout expires. If dynamic member, it will re-assign them right away).

Reply:
Oh, I was not aware of the static membership handling in KIP-848. Thanks for the clarification!

Yes, I agree that consumers that use group management should always send a leave group as KIP-848 describes.

So the comment is wrong. I would propose to just remove the whole comment and give the timer a descriptive name. In general I find inline comments problematic because -- at some point -- they start to lie because they are never properly maintained. Instead I prefer to use meaningful and descriptive names and to extract code that need a comment to methods that either have a meaningful name or a comment on the method, preferring the meaningful name. What I just wrote applies in most cases but not in all, but it definitely applies here.